### PR TITLE
Add per-foot height sensor for terrain-aware gait rewards

### PR DESCRIPTION
**Pull request by kevinzakka**
Foot height sensing now correctly measures vertical clearance above terrain instead of raw ray distances. This fixes feet_clearance, feet_swing_height, and foot_height observations which were computing incorrect values on non-flat terrain.

Adds RayCastSensor.per_frame_height() to compute true vertical heights (frame_z - hit_z) per frame.

### CODE REVIEW

File: src/mjlab/tasks/velocity/mdp/observations.py [hunk 1]
```diff
@@ -42,3 +44,22 @@ def foot_contact_forces(env: ManagerBasedRlEnv, sensor_name: str) -> torch.Tenso
...
+    Tensor of shape [B, F] where F is the number of frames (feet).
+  """
+  sensor: RayCastSensor = env.scene[sensor_name]
+  data = sensor.data
+  F, N = sensor.num_frames, sensor.num_rays_per_frame
+  B = data.distances.shape[0]
+  dists = data.distances.view(B, F, N)
+  miss = dists < 0
+  dists = torch.where(miss, torch.full_like(dists, sensor.cfg.max_distance), dists)
+  return dists.min(dim=-1).values  # [B, F]
```
Comment (mktk1117) [inline src/mjlab/tasks/velocity/mdp/observations.py]: Is this correct? better to double check if there are offsets above. 
Would be better to get the pos of the hit like done in the height scan obs.

File: src/mjlab/tasks/velocity/mdp/observations.py [hunk 2]
```diff
@@ -42,3 +44,22 @@ def foot_contact_forces(env: ManagerBasedRlEnv, sensor_name: str) -> torch.Tenso
   assert sensor_data.force is not None
   forces_flat = sensor_data.force.flatten(start_dim=1)  # [B, N*3]
   return torch.sign(forces_flat) * torch.log1p(torch.abs(forces_flat))
+
+
+def _foot_terrain_height(env: ManagerBasedRlEnv, sensor_name: str) -> torch.Tensor:
```
Comment (mktk1117) [inline src/mjlab/tasks/velocity/mdp/observations.py]: Having this in obs and using in reward makes it a bit hard to follow. 
Either have this in the reward.py or implement this logic as a sensor... ?
Like have some aggregation method (min, max, average median etc)

File: src/mjlab/tasks/velocity/mdp/rewards.py [hunk 3]
```diff
@@ -193,27 +196,27 @@ class feet_swing_height:
   """Penalize deviation from target swing height, evaluated at landing."""
 
   def __init__(self, cfg: RewardTermCfg, env: ManagerBasedRlEnv):
-    self.sensor_name = cfg.params["sensor_name"]
-    self.site_names = cfg.params["asset_cfg"].site_names
+    self.contact_sensor_name = cfg.params["sensor_name"]
```
Comment (louislelay) [inline src/mjlab/tasks/velocity/mdp/rewards.py]: Isn't this a dead assignment?

File: src/mjlab/tasks/velocity/mdp/rewards.py [hunk 4]
```diff
@@ -167,16 +169,17 @@ def feet_air_time(
...
   asset_cfg: SceneEntityCfg = _DEFAULT_ASSET_CFG,
 ) -> torch.Tensor:
   """Penalize deviation from target clearance height, weighted by foot velocity."""
   asset: Entity = env.scene[asset_cfg.name]
-  foot_z = asset.data.site_pos_w[:, asset_cfg.site_ids, 2]  # [B, N]
+  foot_height = _foot_terrain_height(env, height_sensor_name)  # [B, F]
   foot_vel_xy = asset.data.site_lin_vel_w[:, asset_cfg.site_ids, :2]  # [B, N, 2]
   vel_norm = torch.norm(foot_vel_xy, dim=-1)  # [B, N]
-  delta = torch.abs(foot_z - target_height)  # [B, N]
+  delta = torch.abs(foot_height - target_height)  # [B, N]
```
Comment (louislelay) [inline src/mjlab/tasks/velocity/mdp/rewards.py]: NIT: `foot_height` is annotated [B, F] but delta (derived from it) is annotated [B, N]. Since F and N both mean "number of feet" here, should we pick one notation and use it consistently to avoid confusion?

File: src/mjlab/tasks/velocity/velocity_env_cfg.py
Comment (louislelay) [inline src/mjlab/tasks/velocity/velocity_env_cfg.py:1]: In the PR description, you mention:

> The spawn Z range is widened from (0.01, 0.05) to (0.05, 0.10) to reduce initial terrain penetration.

But `pose_range` z is still (0.01, 0.05) in the diff. Was this change dropped intentionally or is it still sitting in your stash?

Comment (kevinzakka) [reply src/mjlab/tasks/velocity/velocity_env_cfg.py:1]: It's something I had but removed because didn't wanna pollute the PR. Updated the PR description though, thanks Louis.

File: src/mjlab/tasks/velocity/mdp/rewards.py (hunk 4 above)
Comment (kevinzakka) [reply src/mjlab/tasks/velocity/mdp/rewards.py]: Fixed!

File: src/mjlab/tasks/velocity/mdp/rewards.py (hunk 3 above)
Comment (kevinzakka) [reply src/mjlab/tasks/velocity/mdp/rewards.py]: Fixed!

File: src/mjlab/tasks/velocity/mdp/observations.py (hunk 2 above)
Comment (kevinzakka) [reply src/mjlab/tasks/velocity/mdp/observations.py]: Implemented as a sensor method: `RayCastSensor.per_frame_height()`. Both obs and rewards now call the sensor directly instead of a private helper in obs.

File: src/mjlab/tasks/velocity/mdp/observations.py (hunk 1 above)
Comment (kevinzakka) [reply src/mjlab/tasks/velocity/mdp/observations.py]: Done.

**Comment by kevinzakka** [comment]
Thanks @mktk1117 and @louislelay for the review!

File: src/mjlab/sensor/raycast_sensor.py
Comment (mktk1117) [inline src/mjlab/sensor/raycast_sensor.py:1]: I feel it's too specific to terrain height to put in to the RayCastSensor. 
If we want to use for other directions (for example spherical scan for obstacle avoidance) we want a ray's distance. 
So it might be better to separate this from the RayCastSensor to keep the ray cast sensor to purely get the ray distance, hit pos etc. And more specific feature to other function, or create a custom sensor inheriting raycastsensor.

Comment (mktk1117) [reply src/mjlab/sensor/raycast_sensor.py:1]: If we make a thin custom sensor inherited from raycast sensor, we can make, different types of sensors which uses raycasting.

MinDistanceSensor (Take min of rays. useful for obstacle avoidance)
TerrainHeightSensor
DepthCameraSensor (we need to project the ray's distance for depth image conversion)
VoxelSensor
(CeilingHeightSensor...?) etc.

Comment (kevinzakka) [reply src/mjlab/sensor/raycast_sensor.py:1]: Fixed.

File: src/mjlab/sensor/terrain_height_sensor.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    )
+
+    reduction = self.cfg.reduction
+    if reduction == "min":
+      reduced = heights.min(dim=-1).values
+    elif reduction == "max":
+      reduced = heights.max(dim=-1).values
+    elif reduction == "mean":
+      reduced = heights.mean(dim=-1)
+    else:
```
Comment (mktk1117) [inline src/mjlab/sensor/terrain_height_sensor.py:80]: sorry, very tiny thing. Maybe add reduction none or raw.

Comment (kevinzakka) [reply src/mjlab/sensor/terrain_height_sensor.py:80]: Done!
